Services/DataSet/classes/class.ilDataSet.php

```diff
@@ -521,25 +518,26 @@ protected function parseObjectExportId(string $a_id, ?string $a_fallback_id = nu
             else {
                 $id = ilObject::_getIdForImportId($a_id);
                 // matching type?
-                if ($id && ilObject::_lookupType($id) == $type) {
+                if ($id && ilObject::_lookupType($id) === $type) {
                     return array("type" => self::EXPORT_ID_ILIAS_REMOTE, "id" => $id);
                 }
                 // not found
                 else {
                     return array("type" => self::EXPORT_ID_ILIAS_REMOTE_INVALID, "id" => $a_fallback_id);
                 }
             }
-        }
-        
-        // external id
-        $id = ilObject::_getIdForImportId($a_id);
-        if ($id) {
-            return array("type" => self::EXPORT_ID, "id" => $id);
         } else {
-            return array("type" => self::EXPORT_ID_INVALID, "id" => $a_fallback_id);
+            // external id
```

Comment:
@pascalseeland @mjansenDatabay  Just a note: It is not necessary to nest this last block into the else part, since all others above will ensure returns. The [EA] inspections even prefer to get rid of all the else branches after ifs which include a return. It says "[EA] Child instructions can be extracted here (clearer intention, lower complexity numbers" for these cases. Actually I am not convinced that this really improves readability of the code. I don't have (not yet) a strong preference either way...

Reply:
@alex40724 

IMO the `Redundant 'else' keyword` warning is **valid**. Every control structure/keyword and nesting which could be simplified should be simplified.

Examples (I suggest ot refactor all of them):

```php
    private function convertInput($a_value) : string
    {
        if (is_array($a_value)) {
            return $a_value[0];
        } else {
            return $a_value;
        }
    }
```

```php
    public function trigger($params)
    {
        if ($this->detection_state == false) {
            $this->setDetectorState(true);
            return true;
        } else {
            return false;
        }
    }
```

```php
    private function getPearAtributeArray() : array
    {
        if ($this->enabledSyncOnLogin()) {
            $mapping = ilLDAPAttributeMapping::_getInstanceByServerId($this->getServerId());
            return array_merge(
                array($this->getUserAttribute()),
                $mapping->getFields(),
                array('dn'),
                ilLDAPRoleAssignmentRules::getAttributeNames($this->getServerId())
            );
        } else {
            return array($this->getUserAttribute());
        }
    }
```

I would also strongly recommed to refactor this kind of of code ...

```php
    // ... a lot of code before this
 
   if ($condition === true) {
       $translation = $this->lng->txt('my_lang_id_a');
   }  else {
       $translation = $this->lng->txt('my_lang_id_b');
   }

    // ... a lot of code after this
```

... to:

```php
    // ... a lot of code before this
 
   $translation = $this->lng->txt('my_lang_id_b');
   if ($condition === true) {
       $translation = $this->lng->txt('my_lang_id_a');
   }

    // ... a lot of code after this
```

And yes, there might be exceptions: If the default statement is `expensive` (in regards to performance), it should not be refactored it this way.

------

But I **don't** suggest to convert `elseif` to `if` branches (maybe only in rare cases),

```php
    public function attemptTransition() : bool
    {
        $eval_function = function ($detectors) {
            return eval($this->evaluation_expression);
        };

        if ($eval_function($this->detectors) === null) {
            return false;
        }
        
        if ($eval_function($this->detectors) === true) {
            $this->executeTransition();
            return true;
        } elseif (method_exists($this, 'executeElseTransition')) {
            $this->executeElseTransition();
            return true;
        }
        return false;
    }
```